resources/datastore.go

```diff
@@ -1,64 +1,74 @@
-package datastore
+package resources
 
 import (
 	"context"
 	"errors"
+	"fmt"
 
 	"github.com/hashicorp/terraform/helper/schema"
 	"github.com/severalnines/terraform-provider-ccx/ccx"
-	chttp "github.com/severalnines/terraform-provider-ccx/http"
-	"github.com/severalnines/terraform-provider-ccx/http/auth"
-	datastoreclient "github.com/severalnines/terraform-provider-ccx/http/datastore-client"
-	"github.com/severalnines/terraform-provider-ccx/terraform"
+	"github.com/severalnines/terraform-provider-ccx/ccx/api"
 )
 
 var (
-	_ ccx.TerraformResource = &Resource{}
+	_ TerraformResource = &Datastore{}
 )
 
-func ToDatastore(d *schema.ResourceData) ccx.Datastore {
+func schemaToDatastore(d *schema.ResourceData) (ccx.Datastore, error) {
 	c := ccx.Datastore{
 		ID:                d.Id(),
-		Name:              terraform.GetString(d, "name"),
-		Size:              terraform.GetInt(d, "size"),
-		DBVendor:          vendorToCCX(terraform.GetString(d, "db_vendor")),
-		DBVersion:         terraform.GetString(d, "db_version"),
-		Type:              terraform.GetString(d, "type"),
-		Tags:              terraform.GetStrings(d, "tags"),
-		CloudProvider:     terraform.GetString(d, "cloud_provider"),
-		CloudRegion:       terraform.GetString(d, "cloud_region"),
-		InstanceSize:      terraform.GetString(d, "instance_size"),
-		VolumeType:        terraform.GetString(d, "volume_type"),
-		VolumeSize:        terraform.GetInt(d, "volume_size"),
-		VolumeIOPS:        terraform.GetInt(d, "volume_iops"),
-		NetworkType:       terraform.GetString(d, "network_type"),
-		HAEnabled:         terraform.GetBool(d, "network_ha_enabled"),
-		VpcUUID:           terraform.GetString(d, "network_vpc_uuid"),
-		AvailabilityZones: terraform.GetStrings(d, "network_az"),
+		Name:              getString(d, "name"),
+		Size:              getInt(d, "size"),
+		DBVendor:          getString(d, "db_vendor"),
+		DBVersion:         getString(d, "db_version"),
+		Type:              getString(d, "type"),
+		Tags:              getStrings(d, "tags"),
+		CloudProvider:     getString(d, "cloud_provider"),
+		CloudRegion:       getString(d, "cloud_region"),
+		InstanceSize:      getString(d, "instance_size"),
+		VolumeType:        getString(d, "volume_type"),
+		VolumeSize:        getInt(d, "volume_size"),
+		VolumeIOPS:        getInt(d, "volume_iops"),
+		NetworkType:       getString(d, "network_type"),
+		HAEnabled:         getBool(d, "network_ha_enabled"),
+		VpcUUID:           getString(d, "network_vpc_uuid"),
+		AvailabilityZones: getStrings(d, "network_az"),
 	}
+
+	dbparams := getMapString(d, "db_params")
+	c.DbParams = dbparams
+
+	firewalls, err := getFirewalls(d, "firewall")
+	if err != nil {
+		return c, err
+	}
+
+	c.FirewallRules = firewalls
+
 	c.Type = defaultType(c.DBVendor, c.Type)
 
-	return c
+	return c, nil
 }
 
-func ToSchema(d *schema.ResourceData, c ccx.Datastore) error {
+func datastoreToSchema(c ccx.Datastore, d *schema.ResourceData) error {
```

Comment:
These names are confusing me. It's not making a schema, right? It's making a ResourceData.

Reply:
Also, we have some functions `XToY` and some `CreateYFromX`, we should probably choose one. I like `YFromX`,